File: 5. VarianceRetention.cpp
// Topic: Calculate variance retained after reducing dimensions using top k eigenvalues
#include <bits/stdc++.h>
using namespace std;

double varianceRetained(const vector<double>& eigenvals, int k)
{
    double total = 0;
    for (double v : eigenvals) total += v;
    double top = 0;
    for (int i = 0; i < k && i < eigenvals.size(); ++i) top += eigenvals[i];
    return top / total;
}

int main()
{
    vector<double> evals = {5.0, 1.5, 0.3};
    cout << "Variance retained by k=1: " << varianceRetained(evals, 1) << '\n';
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(1)

File: 5. VarianceRetention.py
# Topic: Calculate variance retained after reducing dimensions using top k eigenvalues

import numpy as np

def variance_retained(eigenvals, k):
    eigenvals = np.array(eigenvals)
    total = eigenvals.sum()
    return eigenvals[:k].sum() / total

if __name__ == '__main__':
    print('Variance retained by k=1:', variance_retained([5.0,1.5,0.3],1))

# Time Complexity: O(n)
# Space Complexity: O(1)
